### PR TITLE
Adds support for TTL in Memcached State Store

### PR DESCRIPTION
**Pull request by UmutComlekci**
# Description

Adds support for TTL in Memcached State Store

## Issue reference

#306 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [X] Code compiles correctly
* [X] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

## UPDATE

Moving this to 1.4, so will not be in 1.3 as part of #306

### CODE REVIEW

File: state/memcached/memcached.go [hunk 1]
```diff
@@ -15,12 +15,13 @@ import (
...
-	timeout            = "timeout"
+	hosts              			= "hosts"
+	maxIdleConnections 			= "maxIdleConnections"
+	timeout            			= "timeout"
+	ttlInSeconds       			= "ttlInSeconds"
 	// These defaults are already provided by gomemcache
-	defaultMaxIdleConnections = 2
-	defaultTimeout            = 1000 * time.Millisecond
+	defaultMaxIdleConnections 	= 2
+	defaultTimeout            	= 1000 * time.Millisecond
```
Comment (berndverst) [inline state/memcached/memcached.go:24]: Could you undo this whitespace change please?

File: state/memcached/memcached.go [hunk 2]
```diff
@@ -102,10 +103,32 @@ func getMemcachedMetadata(metadata state.Metadata) (*memcachedMetadata, error) {
 	return &meta, nil
 }
 
+func (r *Memcached) parseTTL(req *state.SetRequest) (int, error) {
```
Comment (berndverst) [inline state/memcached/memcached.go:106]: I recommend changing the return type to int32 here instead to be compatible with the Memcache client

File: state/memcached/memcached.go [hunk 3]
```diff
@@ -102,10 +103,32 @@ func getMemcachedMetadata(metadata state.Metadata) (*memcachedMetadata, error) {
...
+		parsedVal, err := strconv.ParseInt(val, 10, 0)
+		if err != nil {
+			return 0, err
+		}
+
+		if parsedVal < 0 {
+			return 0, nil
+		}
+
+		return int(parsedVal), nil
```
Comment (berndverst) [inline state/memcached/memcached.go:117]: return int32 here

File: state/memcached/memcached.go [hunk 4]
```diff
@@ -102,10 +103,32 @@ func getMemcachedMetadata(metadata state.Metadata) (*memcachedMetadata, error) {
...
 func (m *Memcached) setValue(req *state.SetRequest) error {
 	var bt []byte
+	ttl, err := m.parseTTL(req)
+	if err != nil {
+		return fmt.Errorf("failed parse ttl %s: %s", req.Key, err)
+	}
+
 	bt, _ = utils.Marshal(req.Value, m.json.Marshal)
-	err := m.client.Set(&memcache.Item{Key: req.Key, Value: bt})
+	err = m.client.Set(&memcache.Item{Key: req.Key, Value: bt, Expiration: int32(ttl)})
```
Comment (berndverst) [inline state/memcached/memcached.go:131]: remove this cast to int32 in favor of the updated helper function

**Review by berndverst — COMMENTED** [review]
Thanks for helping me implement this in other state stores.

File: state/memcached/memcached.go [hunk 5]
```diff
@@ -15,12 +15,13 @@ import (
...
-	maxIdleConnections = "maxIdleConnections"
-	timeout            = "timeout"
+	hosts              			= "hosts"
+	maxIdleConnections 			= "maxIdleConnections"
+	timeout            			= "timeout"
+	ttlInSeconds       			= "ttlInSeconds"
 	// These defaults are already provided by gomemcache
-	defaultMaxIdleConnections = 2
-	defaultTimeout            = 1000 * time.Millisecond
+	defaultMaxIdleConnections	= 2
```
Comment (UmutComlekci) [inline state/memcached/memcached.go:23]: Actually it's look normal in my editor (VS Code), but in the PR looks like that. I changed encoding but  It's not resolve the problem. Should I need to try other solutions?

Comment (berndverst) [reply state/memcached/memcached.go:23]:
```suggestion
	defaultMaxIdleConnections = 2
```

File: state/memcached/memcached.go [hunk 6]
```diff
@@ -15,12 +15,13 @@ import (
...
+	defaultTimeout				= 1000 * time.Millisecond
```
Comment (berndverst) [inline state/memcached/memcached.go:24]:
```suggestion
	defaultTimeout            = 1000 * time.Millisecond
```

File: state/memcached/memcached.go (hunk 5 above)
Comment (berndverst) [reply state/memcached/memcached.go:23]: You can just accept this and then it should work.

File: state/memcached/memcached.go (hunk 6 above)
Comment (berndverst) [reply state/memcached/memcached.go:24]: You should just be able to accept this to fix it.

File: state/memcached/memcached.go [hunk 7]
```diff
@@ -102,10 +103,32 @@ func getMemcachedMetadata(metadata state.Metadata) (*memcachedMetadata, error) {
 	return &meta, nil
 }
 
+func (r *Memcached) parseTTL(req *state.SetRequest) (int32, error) {
```
Comment (artursouza) [inline state/memcached/memcached.go:106]: Return *int32 and make it a difference if TTL is set vs value is 0.

File: state/memcached/memcached.go [hunk 8]
```diff
@@ -102,10 +103,32 @@ func getMemcachedMetadata(metadata state.Metadata) (*memcachedMetadata, error) {
...
 func (m *Memcached) setValue(req *state.SetRequest) error {
 	var bt []byte
+	ttl, err := m.parseTTL(req)
+	if err != nil {
+		return fmt.Errorf("failed parse ttl %s: %s", req.Key, err)
+	}
+
 	bt, _ = utils.Marshal(req.Value, m.json.Marshal)
-	err := m.client.Set(&memcache.Item{Key: req.Key, Value: bt})
+	err = m.client.Set(&memcache.Item{Key: req.Key, Value: bt, Expiration: ttl})
```
Comment (artursouza) [inline state/memcached/memcached.go:131]: Do not set Expiration if ttl is nil.

**Review by artursouza — CHANGES_REQUESTED** [review]

**Comment by artursouza** [comment]
Moving this to 1.4, so will not be in 1.3 as part of #306

**Comment by berndverst** [comment]
@UmutComlekci to make sure you get credit for the contribution I am going to fork your fork and will patch this up and make a new PR. Today is code freeze for the 1.3 release and I want to make sure this can get in!

**Comment by UmutComlekci** [comment]
> @UmutComlekci to make sure you get credit for the contribution I am going to fork your fork and will patch this up and make a new PR. Today is code freeze for the 1.3 release and I want to make sure this can get in!

Sorry for delay and thank you very much for all your help :)

File: state/memcached/memcached.go (hunk 8 above)
Comment (UmutComlekci) [reply state/memcached/memcached.go:131]: Requested changes completed with #1011

File: state/memcached/memcached.go (hunk 7 above)
Comment (UmutComlekci) [reply state/memcached/memcached.go:106]: Requested changes completed with #1011

**Comment by berndverst** [comment]
@artursouza we should close this PR as it is continued in https://github.com/dapr/components-contrib/pull/1011

**Pull request closed without merging**